docs/_advanced/authorization.md

```diff
@@ -5,41 +5,52 @@ order: 2
 
-* `body` will be provided as the request body of the event.
+The `authorize` option should be set to a function that takes an event source as its input, and should return a Promise for an object containing the authorized credentials. The source contains information about who and where the event is coming from using properties like `teamId` (always available), `userId`, `conversationId`, and `enterpriseId`.
 
-After you app is initialized, the authorization function will be run as an incoming event is passed from the Receiver. This will give opportunity to add the proper token and bot user and app IDs (which are used for the builtin `ignoreSelf` middleware).
+The authorized credentials should also have a few specific properties: `botToken`, `userToken`, `botId` (required for app to ignore messages from itself), and `botUserId`. You can also include any other properties you'd like to make available on the [`context`](#context).
+
+You should always provide either one or both of the `botToken` and `userToken` properties. One of them is necessary to make helpers like `say()` work. If they are both given, then `botToken` will take precedence.
 </div>
 
 ```javascript
+const app = new App({ authorize: authorizeFn, signingSecret: process.env.SLACK_SIGNING_SECRET });
+
 // NOTE: This is for demonstration purposes only.
 // All sensitive data should be stored in a secure database
-let teamInfo = {
-  'E1234A12AB': {
-    'T12345': {
-      botToken: 'xoxb-123abc',
-      botId: 'B1251',
-      botUserId: 'U12385' 
-    }
+// Assuming this app only uses bot tokens, the following object represents a model for storing the credentials as the app is installed into multiple workspaces.
+
+const installations = [
+  {
+    enterpriseId: 'E1234A12AB',
+    teamId: 'T12345',
+    botToken: 'xoxb-123abc',
+    botId: 'B1251',
+    botUserId: 'U12385',
   },
-  'E5678B23CD': {
-    'T77712': {
-      botToken: 'xoxb-102anc',
-      botId: 'B5910',
-      botUserId: 'U1239'
-    }
-  }
-}
+  {
+    teamId: 'T77712',
+    botToken: 'xoxb-102anc',
+    botId: 'B5910',
+    botUserId: 'U1239',
+  },
+];
 
-const authorizeFn = async (source, body) => {
+const authorizeFn = async ({ teamId, enterpriseId }) => {
   // Fetch team info from database. You could also set userToken instead.
   const fetchAuthorizedTeam = new Promise((resolve, reject) => {
-    teamInfo[source.enterpriseId][source.teamId] !== undefined ?
-      Promise.resolve(teamInfo[source.enterpriseId][source.teamId]) :
-      Promise.reject();
+    for (const team in installations) {
+      // Check for matching teamId and enterpriseId in the installations array
+      if ((team.teamId === teamId) && (team.enterpriseId === enterpriseId)) {
+        // This is a match. Use these installaton credentials.
+        Promise.resolve(team);
```

Comment:
sorry, this is a nitpick. can we do something like `return { botToken: team.botToken, botId: team.botId, botUserId: team.botUserId };` instead? first, wrapping in `Promise.resolve()` is unnecessary since we are in an async function. second, i think it might be important to "show our work" and that the naming of the properties is explicit and important (not just rely on the coincidence that the persistence model happens to also use a compatible shape).